### PR TITLE
[Frontend][Runtime] A plugin system for the backend devices

### PR DESCRIPTION
**Pull request by maliasadi**
**Description of the Change:**
- [X] Update the device info communication API. We pass a tuple of (`rtd_name`, `rtd_lib`, `rtd_kwargs`) representing the device name, the shared library path, and other specifications (number of shots, S3, ARN, etc.) from the frontend to the backend. 
- [X] Better encapsulating the runtime utility functions in the frontend by adding `runtime.py`.
- [X] Re-structure the runtime to fully split the backend devices build system from the runtime CAPI and execution context manager. 
- [X] Update the initial dynamic loading support (the custom device interface) in runtime by adding the unique device factory method and simplifying the API.  
- [X] Package all 3 backend devices separately from the runtime and add them all to the new plugin system of Catalyst. Users are now able to dynamically load (custom) devices at runtime.
- [X] Update frontend and backend tests to fully support the plugin system.
- [X] Test the support with Address Sanitizer and `gcov`
- [X] Test the support on macOS and Linux

**Performance:**
The new plugin system of Catalyst is backed by `dlopen` to load libraries at runtime instead of linking them at compile time. This approach may introduce negligible overhead depending on your QJIT workflow. I only observed a slight slowdown when loading multiple devices at runtime inside a JIT-decorated workflow.

Running all frontend tests on a `c5.9xlarge` machine,
Without the plugin system (`backend='lightning.qubit'`):
```
command: make pytest
number of workers: 18 [962 passed]
mean of 5 calls 374.85 sec
```

With the plugin system (`backend='lightning.qubit'`):
```
command: make pytest
number of workers: 18 [962 passed]
mean of 5 calls: 377.22 sec
```

Note that to support the plugin system on CIs, a few frontend tests were skipped. I also skipped those tests in the benchmark. 

[sc-46412]

### CODE REVIEW

File: runtime/Makefile [hunk 1]
```diff
@@ -8,8 +8,9 @@ MK_DIR := $(dir $(MK_ABSPATH))
 RT_BUILD_DIR?=$(MK_DIR)/build
 CMAKE_ARGS?=""
 CODE_COVERAGE?=OFF
-BUILD_TYPE?=Release
+BUILD_TYPE?=RelWithDebInfo
```
Comment (erick-xanadu) [inline runtime/Makefile:11]: Why the change here?

File: runtime/lib/capi/ExecutionContext.hpp [hunk 2]
```diff
@@ -193,55 +213,70 @@ class ExecutionContext final {
...
+            // of the runtime.
+            // Besides, it provides support for libraries added to
+            // the system path.
+            if (rtd_lib == "lightning.qubit") {
+                _device_name = "LightningSimulator";
+#ifdef __linux__
+                rtd_lib = "librtd_lightning.so";
+#elif defined(__APPLE__)
+                rtd_lib = "librtd_lightning.dylib";
+#endif
```
Comment (erick-xanadu) [inline runtime/lib/capi/ExecutionContext.hpp:242]: Can we have different functions here? Example pseudcode

```c
inline char* get_suffix_apple() {
#ifdef defined(_APPLE__)
  return ".dylib";
#else
  return NULL;
#endif
}

inline char* get_suffix_linux() {
#ifdef __linux__
  return ".so";
#else
  return NULL;
#endif
}

char* get_suffix_platform() {
  return get_suffix_linux() | get_suffix_apple();
}
```

File: runtime/lib/capi/ExecutionContext.hpp [hunk 3]
```diff
@@ -72,11 +110,16 @@ class SharedLibraryManager final {
...
         // macOS doesn't support RTLD_DEEPBIND
         _handler = dlopen(filename.c_str(), RTLD_LAZY);
 #else
-        _handler = dlopen(filename.c_str(), RTLD_LAZY | RTLD_DEEPBIND);
+        // Closing the dynamic library of Lightning simulators with dlclose() where
+        // OpenMP directives are in use would raise memory segfaults
+        // Ali: Adding RTLD_NODELETE fixed the issue.
+
+        _handler = dlopen(filename.c_str(), RTLD_LAZY | RTLD_NODELETE);
 #endif
```
Comment (erick-xanadu) [inline runtime/lib/capi/ExecutionContext.hpp:123]: Similar comment to flag below.

**Review by erick-xanadu — CHANGES_REQUESTED** [review]

File: runtime/lib/capi/ExecutionContext.hpp (hunk 2 above)
Comment (maliasadi) [reply runtime/lib/capi/ExecutionContext.hpp:242]: Oh yes, that's the motivation behind the added TODO in code comments.

Comment (maliasadi) [reply runtime/lib/capi/ExecutionContext.hpp:242]: Oh yes, that's the motivation behind the added TODO in code comments a few lines above.

File: runtime/Makefile (hunk 1 above)
Comment (maliasadi) [reply runtime/Makefile:11]: What do you mean? Are you against this change?

File: doc/dev/custom_devices.rst [hunk 4]
```diff
@@ -17,91 +17,95 @@ Additionally, all measurements will always return ``true``.
...
+        return new CustomDevice(status, kwargs); }
 
-where ``CustomDevice()`` is a constructor for your custom device.
-``CustomDevice``'s destructor will be called by the runtime.
+where ``CustomDevice(status, kwargs)`` serves as a constructor for your custom device,
+with `status` representing the initial state of the tape recording system, and `kwargs`
+representing a dictionary of device specifications in string format. The destructor of
+``CustomDevice`` will be automatically called by the runtime.
 
 .. note::
```
Comment (josh146) [inline doc/dev/custom_devices.rst:104]:
```suggestion
.. warning::
```

File: doc/dev/custom_devices.rst [hunk 5]
```diff
@@ -149,11 +153,14 @@ After doing so, Catalyst should be able to interface with your custom device.
 
         @staticmethod
         def get_c_interface():
-            """Location to shared object with C/C++ implementation"""
-            return "full/path/to/libdummy_device.so"
+            """ The tuple of davice name and the location
+                to shared object with C/C++ implementation.
+            """
+
+            return "CustomDevice", "full/path/to/libdummy_device.so"
```
Comment (josh146) [inline doc/dev/custom_devices.rst:160]:
```suggestion
            """Returns a tuple consisting of the device name, and
            the location to the shared object with the C/C++ device implementation.
            """

            return "CustomDevice", "full/path/to/libdummy_device.so"
```

Comment (josh146) [reply doc/dev/custom_devices.rst:160]: @maliasadi it's not entirely clear to me if the path here is relative to some directory?

**Review by josh146 — APPROVED** [review]
Looks good on my end, however I only reviewed the documentation so would still need additional technical review :)

File: doc/dev/custom_devices.rst [hunk 6]
```diff
@@ -17,91 +17,95 @@ Additionally, all measurements will always return ``true``.
...
+    CustomDeviceFactory(bool status, const std::string &kwargs)
+    { 
+        return new CustomDevice(status, kwargs); }
 
-where ``CustomDevice()`` is a constructor for your custom device.
-``CustomDevice``'s destructor will be called by the runtime.
+where ``CustomDevice(status, kwargs)`` serves as a constructor for your custom device,
+with `status` representing the initial state of the tape recording system, and `kwargs`
+representing a dictionary of device specifications in string format. The destructor of
+``CustomDevice`` will be automatically called by the runtime.
```
[the remark this answers is not in the document]
Comment (dime10) [reply doc/dev/custom_devices.rst:102]: Btw, is the tape something that any device can use? Or is it somehow interconnect with Lighting specifically?

File: frontend/catalyst/pennylane_extensions.py [hunk 7]
```diff
@@ -247,8 +206,12 @@ class QJITDevice(qml.QubitDevice):
         "Hamiltonian",
     ]
 
-    def __init__(self, shots=None, wires=None, backend_name=None, backend_kwargs=None):
+    # pylint: disable=too-many-arguments
+    def __init__(
+        self, shots=None, wires=None, backend_name=None, backend_lib=None, backend_kwargs=None
+    ):
         self.backend_name = backend_name if backend_name else "default"
+        self.backend_lib = backend_lib if backend_lib else ""
```
Comment (dime10) [inline frontend/catalyst/pennylane_extensions.py:214]: Are we using this new property anywhere?

**Review by dime10 — COMMENTED** [review]
Only some initial comments from a quick fly by so far, but I'm excited about merging this in 🤩

File: doc/dev/custom_devices.rst (hunk 5 above)
Comment (dime10) [reply doc/dev/custom_devices.rst:160]: Should be absolute. Maybe "absolute/path/to/libdummy_device.so" would be clearer than "full/path/to/libdummy_device.so"?

Comment (josh146) [reply doc/dev/custom_devices.rst:160]: I think mentioning this in the text might be helpful?

Comment (maliasadi) [reply doc/dev/custom_devices.rst:160]: yep, agree! :100:

File: runtime/lib/capi/ExecutionContext.hpp (hunk 3 above)
Comment (maliasadi) [reply runtime/lib/capi/ExecutionContext.hpp]: The code is clear in this method and is not repeated anywhere else in the runtime. Do you still want this suggestion?

File: runtime/lib/capi/ExecutionContext.hpp [hunk 8]
```diff
@@ -193,55 +209,70 @@ class ExecutionContext final {
...
+            // of the runtime.
+            // Besides, it provides support for libraries added to
+            // the system path.
+            if (rtd_lib == "lightning.qubit") {
+                _device_name = "LightningSimulator";
+#ifdef __linux__
+                rtd_lib = "librtd_lightning.so";
+#elif defined(__APPLE__)
+                rtd_lib = "librtd_lightning.dylib";
+#endif
```
[the remark this answers is not in the document]
Comment (maliasadi) [reply runtime/lib/capi/ExecutionContext.hpp]: Hehe.. there was a big TODO a couple of lines above! ;) I was trying to get C++ tests working with minimal amount of code changes here and there.

The `rtd_lib` isn't necessarily required to be an absolute path to the library. If you call `dlopen` without providing an absolute path, the linker searches for the lib in the system's default search paths. This also considers paths specified in the `LD_LIBRARY_PATH` env variable. In fact, if you set `LD_LIBRARY_PATH`, the dynamic linker will use this path as the default library search paths.

I've added these PL to Runtime device names conversions to extend the support. This also helped me to check all C++ unit tests in the runtime without changing anything specifically.

File: doc/dev/custom_devices.rst (hunk 6 above)
[the remark this answers is not in the document]
Comment (maliasadi) [reply doc/dev/custom_devices.rst:102]: Good points! let me update the guideline.

File: frontend/catalyst/pennylane_extensions.py (hunk 7 above)
Comment (maliasadi) [reply frontend/catalyst/pennylane_extensions.py:214]: not precisely! As the `backend_lib` variable must always initialize with or without an absolute path to the shared library of the backend device.

File: runtime/lib/capi/ExecutionContext.hpp (hunk 2 above)
Comment (erick-xanadu) [reply runtime/lib/capi/ExecutionContext.hpp]: I still see the TODO (?) Have you pushed? It is a minor point, I just really dislike preprocessor macros in the middle of functions and prefer functions that just compile away to no-ops.

Comment (maliasadi) [reply runtime/lib/capi/ExecutionContext.hpp]: There shouldn't be any TODO :thinking: You may want to referesh your browser to see the latest changes in GH.

Me too. I also don't like preprocessors in the middle of functions. In runtime, we don't rely on these pp macros in too many places; they're used in short methods and mostly in `ExecutionContext.hpp`.

File: doc/dev/custom_devices.rst (hunk 6 above)
[the remark this answers is not in the document]
Comment (maliasadi) [reply doc/dev/custom_devices.rst]: Fixed [266cf33](https://github.com/PennyLaneAI/catalyst/pull/343/commits/266cf33edf1975967fd9a2a42ee28a5647c61300). @dime10 lmk what you think and if you're happy with the state of this PR.

File: doc/dev/custom_devices.rst [hunk 9]
```diff
@@ -17,91 +17,95 @@ Additionally, all measurements will always return ``true``.
...
-            DummyDevice &operator=(const QuantumDevice &) = delete;
-            DummyDevice(const DummyDevice &) = delete;
-            DummyDevice(DummyDevice &&) = delete;
-            DummyDevice &operator=(QuantumDevice &&) = delete;
-
-            virtual std::string getName(void) { return "DummyDevice"; }
-
-            auto AllocateQubit() -> QubitIdType { return 0; }
-            virtual auto AllocateQubits(size_t num_qubits) -> std::vector<QubitIdType>
+        struct CustomDevice final : public Catalyst::Runtime::QuantumDevice {
```
[the remark this answers is not in the document]
Comment (dime10) [reply doc/dev/custom_devices.rst:20]: Thanks, these are fair arguments 👍

File: runtime/lib/capi/ExecutionContext.hpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (dime10) [reply runtime/lib/capi/ExecutionContext.hpp]: > The rtd_lib isn't necessarily required to be an absolute path to the library. If you call dlopen without providing an absolute path, the linker searches for the lib in the system's default search paths. This also considers paths specified in the LD_LIBRARY_PATH env variable. In fact, if you set LD_LIBRARY_PATH, the dynamic linker will use this path as the default library search paths.

Hmm, but why is that behaviour we want? It introduces additional potential for errors (e.g. importing a different library with conflicting name, not finding a library, mixing up os-specific file extensions, ...). Making it a hard requirement for a device plugin to provide the exact path to its implementation lib seems safer and simpler to me.

File: frontend/catalyst/pennylane_extensions.py (hunk 7 above)
Comment (dime10) [reply frontend/catalyst/pennylane_extensions.py:214]: Where is the property used?

**Review by erick-xanadu — APPROVED** [review]

File: runtime/lib/capi/ExecutionContext.hpp (hunk 3 above)
Comment (erick-xanadu) [reply runtime/lib/capi/ExecutionContext.hpp]: I do have a preference to avoid preprocessor macros anywhere. But I understand that it is just a preference. The example I showed, I think, is a lot easier to read than having two functionalities (potentially 3, when later we have to support windows). I'll leave this as your choice whether you disagree or not.

File: frontend/catalyst/pennylane_extensions.py (hunk 7 above)
Comment (maliasadi) [reply frontend/catalyst/pennylane_extensions.py:214]: This property contains the name of the shared library (that's different from the name of the device which stores in `backend_name`). All these backend info will be traced and hardcoded to the generated JAXPR by the `qdevice` primitive. The runtime uses the value of `backend_lib` to link to the library and `backend_name` to resolve the symbol of the factory method.

Comment (dime10) [reply frontend/catalyst/pennylane_extensions.py:214]: Ah got it, thanks!

File: runtime/lib/capi/ExecutionContext.hpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (dime10) [reply runtime/lib/capi/ExecutionContext.hpp]: Ok I see the point about making it easier to run tests by maintaining backwards compatibility for specifying a device. Let's add a TODO to deprecate this behaviour in the future :)

File: doc/dev/custom_devices.rst [hunk 10]
```diff
@@ -17,91 +17,116 @@ Additionally, all measurements will always return ``true``.
...
 
     extern "C" Catalyst::Runtime::QuantumDevice*
-    getCustomDevice() { return new CustomDevice(); }
+    CustomDeviceFactory(bool status, const std::string &kwargs)
+    {
+        return new CustomDevice(status, kwargs); }
+
+The use of the term ``Factory`` in ``CustomDeviceFactory`` is a runtime convention that combines the name
+of the device class, in this case, ``CustomDevice``, with the suffix ``Factory``. This signifies that the
+primary purpose of the method is to act as a factory for creating instances of the ``CustomDevice`` class.
```
Comment (dime10) [inline doc/dev/custom_devices.rst:102]:
```suggestion
The entry point function acts as a factory method for the device class. Note that a plugin library
may also provide several factory methods in case it packages multiple devices into the same
library. However, it is important that the device identifier be unique, as best as possible, to
avoid clashes with other plugins. 
```

File: doc/dev/custom_devices.rst [hunk 11]
```diff
@@ -17,91 +17,116 @@ Additionally, all measurements will always return ``true``.
...
+
+.. warning::
+
+    The runtime exclusively supports factory methods with the name of the device class followed by ``Factory``.
+    Please ensure that your method adheres to this naming convention.
```
Comment (dime10) [inline doc/dev/custom_devices.rst:107]:
```suggestion
Importantly, the ``<DeviceIdentifier>`` string in the entry point function needs to match
exactly what is supplied to the ``__quantum__rt__device("rtd_name", "<DeviceIdentifier>")``
runtime instruction in compiled user programs, or what is returned from the ``get_c_interface``
function when integrating the device into a PennyLane plugin (see section further down).
```

File: doc/dev/custom_devices.rst [hunk 12]
```diff
@@ -17,91 +17,116 @@ Additionally, all measurements will always return ``true``.
...
 
-where ``CustomDevice()`` is a constructor for your custom device.
-``CustomDevice``'s destructor will be called by the runtime.
+``CustomDevice(status, kwargs)`` serves as a constructor for your custom device, with
```
Comment (dime10) [inline doc/dev/custom_devices.rst:109]:
```suggestion
Additionally, the entry point function must accept two arguments defined as below, which
can be forwarded to the device constructor (e.g. ``CustomDevice(status, kwargs)``) if
appropriate:
```

File: doc/dev/custom_devices.rst [hunk 13]
```diff
@@ -17,91 +17,116 @@ Additionally, all measurements will always return ``true``.
...
 
-.. note::
+- ``status``: A boolean to control the initial value of the tape recording system in the runtime.
+  If ``true``, this will cache the entire operations and observables in a program.
+  This may be used to compute gradient of a quantum circuit with taking advantage of gradient
+  methods provided by backend devices if supported.
```
Comment (dime10) [inline doc/dev/custom_devices.rst:114]: You mentioned that the tape recording system can be used by any device, but here it looks like the device still needs to keep track of the recording state?

File: doc/dev/custom_devices.rst [hunk 14]
```diff
@@ -17,91 +17,116 @@ Additionally, all measurements will always return ``true``.
...
+
+- ``kwargs``: A string that represents a Python dictionary of device specifications in string format.
+  This may contain information about the device type and a number of shots:
+  ``{'device_type': 'lightning.qubit', 'shots': 0}``.
```
Comment (dime10) [inline doc/dev/custom_devices.rst:118]:
```suggestion
- ``kwargs``: A string of device specifications and options, represented in Python dictionary format.
  This may contain information such as device type and a number of shots:
  ``{'device_type': 'lightning.qubit', 'shots': 0}``.
```

Comment (dime10) [reply doc/dev/custom_devices.rst:118]: What is the purpose of `device_type` here?

File: doc/dev/custom_devices.rst [hunk 15]
```diff
@@ -17,91 +17,116 @@ Additionally, all measurements will always return ``true``.
...
+
+These parameters are automatically initialized in the frontend using the device specifications specified
+by users via :func:`qml.device() <pennylane.device>`.
```
Comment (dime10) [inline doc/dev/custom_devices.rst:121]:
```suggestion
Note that these parameters are automatically initialized in the frontend if the library is provided as a
PennyLane plugin device (see :func:`qml.device() <pennylane.device>`).
```

File: doc/dev/custom_devices.rst [hunk 16]
```diff
@@ -132,14 +157,14 @@ After doing so, Catalyst should be able to interface with your custom device.
```
Comment (dime10) [inline doc/dev/custom_devices.rst:157]:
```suggestion

.. note::

    The first result of ``get_c_interface`` needs to to match the ``<DeviceIdentifier>``
    as described in the first section.

```

**Review by dime10 — APPROVED** [review]
Great work 🥳  I just left some improvements to the documentation to be more specific and that should hopefully make it very clear how to create a custom device for third parties.

File: doc/dev/custom_devices.rst (hunk 14 above)
Comment (erick-xanadu) [reply doc/dev/custom_devices.rst:118]: I would like to preserve `device_type`. I at least like it for the toml files. Right now the names are `lightning.qubit.toml` and that is f`{device_type}.toml`.

File: doc/dev/custom_devices.rst (hunk 13 above)
Comment (dime10) [reply doc/dev/custom_devices.rst:114]: I'm wondering if we need to either explain this component more, since the plugin developer is required to accept this parameter, or if we'd like to remove it somehow?

File: doc/dev/custom_devices.rst (hunk 14 above)
Comment (dime10) [reply doc/dev/custom_devices.rst:118]: If `device_type` is a holdover from before, or specific to only one device, we might want to use a more general example of a parameter a device might accept. But the important part is that this is not a specification, and is just showing that devices can receive arbitrary information from the program in string format.

Comment (maliasadi) [reply doc/dev/custom_devices.rst:118]: This is technically needed for `openqasm` devices to declare the type of the device (remote or local). I re-scoped this to be only appeared for `openqasm` devices (and not for lightning simulators).

File: doc/dev/custom_devices.rst (hunk 11 above)
Comment (josh146) [reply doc/dev/custom_devices.rst:107]: @dime10 curious what your reason is here for removing the `.. warning::` admonition? No strong feelings either way, but I often find that admonitions, if not overused, allow important information to 'stand out' better. Especially in reference documentation, that is often skim read

Comment (dime10) [reply doc/dev/custom_devices.rst:107]: Oh I didn't mean for the change to specifically remove it, I just rewrote the instructions into flowing text as follows:

> In addition to implementing the ``QuantumDevice`` class, one must implement an entry point for the
> device library with the name ``<DeviceIdentifier>Factory``, where ``<DeviceIdentifier>`` is used to
> unique the entry point symbol. As an example, we use the identifier ``CustomDevice``:
>
> ```py
> extern "C" Catalyst::Runtime::QuantumDevice*
> CustomDeviceFactory(bool status, const std::string &kwargs) {
>    return new CustomDevice(status, kwargs);
> }
> ```
>
> The entry point function acts as a factory method for the device class. Note that a plugin library
> may also provide several factory methods in case it packages multiple devices into the same
> library. However, it is important that the device identifier be unique, as best as possible, to
> avoid clashes with other plugins. 
>
> Importantly, the ``<DeviceIdentifier>`` string in the entry point function needs to match
> exactly what is supplied to the ``__quantum__rt__device("rtd_name", "<DeviceIdentifier>")``
> runtime instruction in compiled user programs, or what is returned from the ``get_c_interface``
> function when integrating the device into a PennyLane plugin (see section further down).
>
> Additionally, the entry point function must accept two arguments defined as below, which
> can be forwarded to the device constructor (e.g. ``CustomDevice(status, kwargs)``) if
> appropriate:
> - ``status``: A boolean to control the initial value of the tape recording system in the runtime.
>  If ``true``, this will cache the entire operations and observables in a program.
>  This may be used to compute gradient of a quantum circuit with taking advantage of gradient
>  methods provided by backend devices if supported.
> - ``kwargs``: A string of device specifications and options, represented in Python dictionary format.
>  This may contain information such as device type and a number of shots:
>  ``{'device_type': 'lightning.qubit', 'shots': 0}``.

All of the parts are equally important I think if you want to implement a custom device, but let me know if you think a warning box would still make sense somewhere :)

File: doc/dev/custom_devices.rst (hunk 13 above)
Comment (maliasadi) [reply doc/dev/custom_devices.rst]: Giving it some thoughts and this is not required to be part of the constructor anymore as the `StartTapeRecording` and `StopTapeRecording` methods in the QuantumDevice interface properly satisfies the  objectives of this `status` arg in the constructor. I removed this and updated the docs.

File: doc/dev/custom_devices.rst (hunk 15 above)
Comment (maliasadi) [reply doc/dev/custom_devices.rst]: Updated [e425745](https://github.com/PennyLaneAI/catalyst/pull/343/commits/e42574552ea513430058dec8fee1bb6392071720)

File: doc/dev/custom_devices.rst (hunk 10 above)
Comment (maliasadi) [reply doc/dev/custom_devices.rst]: Updated with some changes [e425745](https://github.com/PennyLaneAI/catalyst/pull/343/commits/e42574552ea513430058dec8fee1bb6392071720)

File: doc/dev/custom_devices.rst [hunk 17]
```diff
@@ -17,91 +17,116 @@ Additionally, all measurements will always return ``true``.
...
             {
             }
 
-            virtual void Gradient(std::vector<DataView<double, 1>> &, const std::vector<size_t> &) {}
+            void Gradient(std::vector<DataView<double, 1>> &, const std::vector<size_t> &) override {}
         };
 
-In addition to implementing the ``QuantumDevice`` class, one must implement the following method:
+In addition to implementing the ``QuantumDevice`` class, one must implement the following method
+with the name ``CustomDevice`` + ``Factory``:
```
[the remark this answers is not in the document]
Comment (maliasadi) [reply doc/dev/custom_devices.rst]: I like `<DeviceIdentifier>` way of describing the API :+1: 
Updated this in the entire guideline [e425745](https://github.com/PennyLaneAI/catalyst/pull/343/commits/e42574552ea513430058dec8fee1bb6392071720)

File: runtime/lib/capi/RuntimeCAPI.cpp [hunk 18]
```diff
@@ -137,17 +137,24 @@ void __quantum__rt__device(int8_t *spec, int8_t *value)
...
     }
-    else if (args[0] == "backend") {
+    else if (args[0] == "rtd_name") {
+        Catalyst::Runtime::CTX->setDeviceName(args[1]);
+        return;
+    }
+    else if (args[0] == "rtd_lib") {
         RT_FAIL_IF(!Catalyst::Runtime::CTX->initDevice(args[1]),
                    "Failed initialization of the backend device");
+        if (Catalyst::Runtime::CTX->getDeviceRecorderStatus()) {
```
Comment (maliasadi) [inline runtime/lib/capi/RuntimeCAPI.cpp:151]: @dime10 That's the update required after removing `status` from the `QuantumDevice` constructor.

Comment (dime10) [reply runtime/lib/capi/RuntimeCAPI.cpp:151]: Nice, even better 💯

File: doc/dev/custom_devices.rst [hunk 19]
```diff
@@ -17,91 +17,100 @@ Additionally, all measurements will always return ``true``.
...
-.. note::
+Importantly, the ``<DeviceIdentifier>`` string in the entry point function needs to match
+exactly what is supplied to the ``__quantum__rt__device("rtd_name", "<DeviceIdentifier>")``
+runtime instruction in compiled user programs, or what is returned from the ``get_c_interface``
+function when integrating the device into a PennyLane plugin. Please see the "Integration with
+Python devices" section further down for details.
+
+``CustomDevice(kwargs)`` serves as a constructor for your custom device, with ``kwargs``
+as a string of device specifications and options, represented in Python dictionary format.
+This may contain information such as device type and a number of shots: ``{'shots': 1000}``.
```
Comment (dime10) [inline doc/dev/custom_devices.rst:102]: Updated version is much clearer, I think we just need to remove `device type` from since there is no context for what it could mean:
```suggestion
An example could bet the number of shots or circuit executions, encoded as follow the following string: ``"{'shots': 1000}"``.
```
(not sure about the quotes or not show the entire thing is a string)

Comment (maliasadi) [reply doc/dev/custom_devices.rst:102]: Sure, I think adding quotes makes more sense as this highlights the fact that this is passed to device as a string.

File: doc/dev/custom_devices.rst (hunk 11 above)
Comment (josh146) [reply doc/dev/custom_devices.rst]: all good! Was just curious if there was an dislike-of-admonitions reason for it lol

File: doc/dev/custom_devices.rst (hunk 19 above)
Comment (maliasadi) [reply doc/dev/custom_devices.rst]: Thanks! Addressed [6db724c](https://github.com/PennyLaneAI/catalyst/pull/343/commits/6db724cdca8959e77ba934db8aec6e21c5c1f055) and [ab5c257](https://github.com/PennyLaneAI/catalyst/pull/343/commits/ab5c25771f948f00f3ecd0716188fab57553ec3b) with some other fixes to the guideline.

File: doc/dev/custom_devices.rst (hunk 16 above)
Comment (maliasadi) [reply doc/dev/custom_devices.rst:140]: Fixed along with other updates [e425745](https://github.com/PennyLaneAI/catalyst/pull/343/commits/e42574552ea513430058dec8fee1bb6392071720)
